### PR TITLE
fix/tests: fix an issue whereby peers' states were not updated

### PR DESCRIPTION
**Pull request by Fraser999**
Upon removing a peer in the test network, we failed to recheck existing peers which were transitioning to Added or Removed to see if the new lower number of members meant that a supermajority was valid now.

Rather than persist with the `PeerRemovalGuard` which effectively represented a "God view" of the network in terms of the joining and leaving states of the peers, there is now a network view held as a
member of the `Peer` struct.

This avoids synchronisation problems and allows better encapsulation of `Peer` by tying the transition between states into `Peer` methods.

This also moves the `PeerStatuses` struct unchanged into its own file to avoid cluttering peer.rs.

### CODE REVIEW

File: src/dev_utils/network.rs [hunk 1]
```diff
@@ -401,15 +412,15 @@ impl Network {
         Ok(())
     }
 
-    /// Checks if the blocks are only signed by valid voters
+    /// Checks if the blocks are only signed by valid voters.
     fn check_blocks_signatories(&self) -> Result<(), ConsensusError> {
-        let blocks = self.active_peers().next().unwrap().blocks();
+        let blocks = unwrap!(self.active_peers().next()).blocks();
```
Comment (jeanphilippeD) [inline src/dev_utils/network.rs]: Why the change? does it do the same thing?

File: src/dev_utils/network.rs [hunk 2]
```diff
@@ -505,216 +514,82 @@ impl Network {
...
-                }
                 self.peers = peers;
                 self.genesis = genesis_group;
-                // do a full reset while we're at it
+                // Do a full reset while we're at it.
                 self.msg_queue.clear();
             }
-            ScheduleEvent::AddPeer(peer) => {
-                let current_peers = self.active_peers().map(|peer| peer.id.clone()).collect();
+            ScheduleEvent::AddPeer(peer_id) => {
```
Comment (jeanphilippeD) [inline src/dev_utils/network.rs]: There are a lot nice small refactoring like this.
It would be good if they were all in a refactoring commit because it adds a lot of noise to the diff.

File: src/dev_utils/network.rs [hunk 3]
```diff
@@ -505,216 +514,82 @@ impl Network {
...
-                        }
-                    }
-
+                for peer_id in self.running_peers_ids() {
+                    self.peer_mut(&peer_id).make_votes();
+                    self.handle_messages(&peer_id, step);
+                    self.peer_mut(&peer_id).poll_all();
+                    self.check_unexpected_accusations(&peer_id)?;
+                }
+                Peer::update_network_views(&mut self.peers);
```
Comment (jeanphilippeD) [inline src/dev_utils/network.rs]: Same comment, but for bigger refactoring it is even more important: Quite difficult to see what changed in the functionality of update_network_views, but would be trivial if we split the change in 2 steps.

File: src/dev_utils/peer.rs [hunk 4]
```diff
@@ -11,27 +11,58 @@ use super::Observation;
...
+    is_more_than_two_thirds, ConsensusMode, Malice, Observation as ParsecObservation,
+};
 use crate::parsec::Parsec;
-use rand::Rng;
 use std::collections::{BTreeMap, BTreeSet};
 use std::fmt::{self, Debug, Formatter};
 
+/// This represents the peer's own view of its current status.
+///
+/// A new peer will start as `Pending` and transition to `Active` once `Peer::poll()` yields a block
```
Comment (jeanphilippeD) [inline src/dev_utils/peer.rs]: I'm happy with this comment, it is really helpful.
I wonder if it would be clearer to describe this small state machine with ASCII art?

File: src/dev_utils/peer.rs [hunk 5]
```diff
@@ -11,27 +11,58 @@ use super::Observation;
...
-    Active,
     Pending,
+    Active,
     Removed,
     Failed,
 }
 
+/// This represents the network's view of a peer's state.
+///
+/// A new peer will start as `Joining`.  While in this state, its own status may or may not
```
Comment (jeanphilippeD) [inline src/dev_utils/peer.rs]: Same

File: src/dev_utils/peer.rs [hunk 6]
```diff
@@ -97,175 +137,152 @@ impl Peer {
...
+        let mut removed_counts = BTreeMap::new();
+        let mut running_peers_count = 0;
+        for peer in all_peers
+            .values()
+            .filter(|peer| peer.network_view == NetworkView::Joined)
+        {
+            for added_peer_id in &peer.added_peers_ids {
+                let count = added_counts.entry(added_peer_id.clone()).or_insert(0);
+                *count += 1;
+            }
```
Comment (jeanphilippeD) [inline src/dev_utils/peer.rs]: I'm not super keen on for within for.

Maybe:
```
let add_count = |new_ids, counts| {
            for id in new_ids {
                let count = counts.entry(id.clone()).or_insert(0);
                *count += 1;
            }
};
...
   add_count( &peer.added_peers_ids, &mut added_counts);
   add_count( &peer.removed_peers_ids, &mut removed_counts);
```


Also maybe in addition can avoid loop for updating ids with something like:
```
let add_count = |new_ids, counts| {
    new_ids.map( |id| {
        counts.entry(id.clone())
           .and_modify(|c| { *c += 1 })
           .or_insert(1);
    } );
};
```

File: src/dev_utils/peer_statuses.rs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::PeerStatus;
+use crate::mock::PeerId;
+use rand::Rng;
+use std::collections::{BTreeMap, BTreeSet};
+
+pub struct PeerStatuses {
```
Comment (jeanphilippeD) [inline src/dev_utils/peer_statuses.rs]: Same, this move/refactor should probably have been done in a separate commit.
I cannot really see what changed.

File: src/dev_utils/network.rs (hunk 1 above)
Comment (Fraser999) [reply src/dev_utils/network.rs]: The macro provides a much more useful error message.  It's been a policy to prefer the macro for quite a while now: https://github.com/maidsafe/QA/blob/master/Documentation/Rust%20Style.md#unwrap

File: src/dev_utils/network.rs (hunk 2 above)
Comment (Fraser999) [reply src/dev_utils/network.rs]: I agree for larger changes (e.g. if rustfmt changes a bunch of files), but for very few, trivial changes like this, I feel it slows down the author more than the reviewers to burst a PR into more commits.

File: src/dev_utils/network.rs (hunk 3 above)
Comment (Fraser999) [reply src/dev_utils/network.rs]: I don't feel this falls into the same category as the previous comment.  It's not a trivial renaming of a variable or function; rather it's a key element of the refactoring effort.

File: src/dev_utils/peer.rs (hunk 4 above)
Comment (Fraser999) [reply src/dev_utils/peer.rs]: I'd be happy to add that, but not quite sure what you had in mind.  The better we can document these states, the easier future debugging will be!

File: src/dev_utils/network.rs (hunk 2 above)
Comment (jeanphilippeD) [reply src/dev_utils/network.rs]: I generally think this is a tooling issue.
Splitting a commit in multiple parts should be a trivial operation (5/15 minutes for a change like this + run minimal testing), so the author's time should not be much of a consideration here.

I would agree if the rest of the code change was small, but since the code change big as well, removing anything not relevant is really important so the review can be effective (not only faster).

Also, help keep the history clean so when we go back we can understand what it does.

File: src/dev_utils/peer.rs (hunk 6 above)
Comment (Fraser999) [reply src/dev_utils/peer.rs]: Well, I like the deduplication aspect of the former option you provided (I considered something like this but rejected the idea as it's a fairly basic two-line function).  I guess I'd still view that as a `for` within `for`, but then I don't have an issue with nesting `for`s, unless the total nested depth gets big (but that applies to any code, not just `for`s).

I can't say I'm a fan of the second option; I think it's mildly harder to read.

File: src/dev_utils/peer_statuses.rs (hunk 7 above)
Comment (Fraser999) [reply src/dev_utils/peer_statuses.rs]: I mentioned in the PR comment that nothing changed.

File: src/dev_utils/peer.rs (hunk 4 above)
Comment (jeanphilippeD) [reply src/dev_utils/peer.rs]: I had something like that in mind. But mostly putting it to show the important info and transition.

```
/// Peer Start
/// |-> `Pending`
///      |
///    Peer::poll()` yields a block)
///      |->`Active`
///          |
///        Need to kill peer
///          |-> `Removed`
///          |-> `Failed`
```

File: src/dev_utils/peer.rs [hunk 8]
```diff
@@ -11,27 +11,58 @@ use super::Observation;
...
     Removed,
     Failed,
 }
 
+/// This represents the network's view of a peer's state.
+///
+/// A new peer will start as `Joining`.  While in this state, its own status may or may not
+/// transition from `PeerStatus::Pending` to `PeerStatus::Active`.  Once a supermajority of
+/// currently `Joined` peers have reached consensus on this peer being added, its network view will
+/// change to `Joined`.
```
Comment (maqi) [inline src/dev_utils/peer.rs]: It might be worth to mention that while a peer is in NetworkView::Joining, it won't be allowed to be removed.

File: src/dev_utils/network.rs (hunk 2 above)
Comment (Fraser999) [reply src/dev_utils/network.rs]: I disagree in this particular case as the cleanups are so small and unobtrusive.  However, I can accept that I'm in the minority here.

File: src/dev_utils/peer.rs (hunk 4 above)
Comment (Fraser999) [reply src/dev_utils/peer.rs]: Added in a fixup commit.

File: src/dev_utils/peer.rs (hunk 8 above)
Comment (Fraser999) [reply src/dev_utils/peer.rs]: That's a good point.  I think we can actually enforce that in the code too by checking the peer's network state when calling the `mark_as_...` methods.

File: src/dev_utils/peer.rs [hunk 9]
```diff
@@ -40,9 +50,22 @@ pub enum PeerStatus {
...
+/// will change to `Left`.
+///
+/// Peer start
+///     |-> `Joining`
+///     |
+///     Supermajority of `Joined` peers see `Add` for this one
+///         |-> `Joined`
+///         |
+///         Any peer votes to remove this one OR this peer is killed
+///             |-> `Leaving`
```
Comment (maqi) [inline src/dev_utils/peer.rs]: In case we will need to similuate the scenario that a malicious node voted to `remove` a peer, will this `any peer votes to remove` triggers an incorrect updated in the network view?

Comment (Fraser999) [reply src/dev_utils/peer.rs]: Well, it should trigger a panic hopefully.  But I don't think we should be catering to that yet, since all our malice scenarios so far are being created by parsing dot files and/or passing malicious messages rather than running actual malicious test nodes.

File: src/dev_utils/peer.rs [hunk 10]
```diff
@@ -97,175 +164,154 @@ impl Peer {
...
     }
 
-    /// Randomly chooses a peer to remove.
-    pub fn remove_random_peer<R: Rng>(&mut self, rng: &mut R, min_active: usize) -> Option<PeerId> {
-        let name = self.choose_name_to_remove(rng);
-
-        let mut active_peers = self.num_active_peers();
-        let mut failed_peers = self.num_failed_peers();
+    // Consider only the peers which the network views as being `Joined`, and check to see if a
+    // supermajority of these have polled blocks which allow the network to update its view of any
```
Comment (jeanphilippeD) [inline src/dev_utils/peer.rs]: Why supermajority?
Why not all, or 50%?

File: src/dev_utils/peer.rs [hunk 11]
```diff
@@ -97,175 +164,154 @@ impl Peer {
...
+    // `Joining` or `Leaving` peers, transitioning these to `Joined` or `Left` respectively.
+    pub fn update_network_views(all_peers: &mut BTreeMap<PeerId, Peer>) {
+        let mut added_counts = BTreeMap::new();
+        let mut removed_counts = BTreeMap::new();
+        let mut running_peers_count = 0;
```
Comment (jeanphilippeD) [inline src/dev_utils/peer.rs]: Should these declaration be moved just above where they are used?

File: src/dev_utils/peer.rs [hunk 12]
```diff
@@ -97,175 +164,154 @@ impl Peer {
...
     }
 
-    pub fn fail_peer(&mut self, peer: &PeerId) {
-        let status = unwrap!(self.statuses.get_mut(peer));
-        *status = PeerStatus::Failed;
+    pub fn is_active_and_has_block(&self, payload: &Observation) -> bool {
+        if self.status != PeerStatus::Active {
+            return false;
+        }
+        self.blocks.iter().any(|block| block.payload() == payload)
```
Comment (jeanphilippeD) [inline src/dev_utils/peer.rs]: self.status == PeerStatus::Active &&
   self.blocks.iter().any(|block| block.payload() == payload)

File: src/dev_utils/network.rs [hunk 13]
```diff
@@ -134,20 +134,33 @@ impl Network {
...
             .values()
-            .filter(move |&peer| peer.status == status)
+            .filter(|peer| peer.status() == PeerStatus::Active)
     }
 
-    fn active_peers(&self) -> impl Iterator<Item = &Peer> {
-        self.peers_with_status(PeerStatus::Active)
+    /// Returns the IDs of peers which consider themselves to be still running correctly, i.e. those
+    /// with `PeerStatus::Pending/Active`.
+    fn running_peers_ids(&self) -> Vec<PeerId> {
```
Comment (jeanphilippeD) [inline src/dev_utils/network.rs]: /// Returns IDs of peer thinking they are running.

Not sure the comment really help though.
Also PeerStatus::Pending/Active`. is an implementation detail of is_running so it is liable to becoming out of date if is_running is changed and should not be mentioned here.

Comment (Fraser999) [reply src/dev_utils/network.rs]: I take your point about about `PeerStatus::Pending/Active` being an implementation detail, but disagree with the rest of your comment.

I'll update the code comment to refer to `is_running()`.

File: src/dev_utils/peer.rs [hunk 14]
```diff
@@ -97,175 +164,154 @@ impl Peer {
...
-        self.peers_by_status(|s| *s == PeerStatus::Active || *s == PeerStatus::Pending)
-            .map(|(id, _)| id)
-    }
-
-    /// Returns an iterator through the list of inactive peers (removed and failed).
-    pub fn inactive_peers(&self) -> impl Iterator<Item = &PeerId> {
-        self.peers_by_status(|s| *s == PeerStatus::Removed || *s == PeerStatus::Failed)
-            .map(|(id, _)| id)
+    /// Sets the node's own status to `Removed` and the network's view of it to `Leaving`.  Panics
+    /// if the node in question isn't yet viewed by the network as being `Joined`.
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply src/dev_utils/peer.rs]: I can revise the comment to something similar to what you suggest:
```
/// Update status and network view when peer is removed from the the network.
```
I don't see that as much of an improvement however, since I'd still keep the part about panicking if not joined.

Another option is to remove the comments for `mark_as_resolved()` and `mark_as_failed()` if others feel they're unhelpful.

File: src/dev_utils/peer.rs [hunk 15]
```diff
@@ -97,175 +164,154 @@ impl Peer {
...
+    pub fn mark_as_removed(&mut self) {
+        if self.status == PeerStatus::Failed {
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply src/dev_utils/peer.rs]: This behaviour is part of `self.mark_as_leaving();`

File: src/dev_utils/peer.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (jeanphilippeD) [reply src/dev_utils/peer.rs]: Maybe the panic part could be more concise and says:
`/// Panic if self network_view is Joining.`

Or:
`/// Expect self network_view is not 'Joining'.`

File: src/dev_utils/peer.rs [hunk 16]
```diff
@@ -97,175 +164,154 @@ impl Peer {
...
-    /// Randomly chooses a peer to remove.
-    pub fn remove_random_peer<R: Rng>(&mut self, rng: &mut R, min_active: usize) -> Option<PeerId> {
-        let name = self.choose_name_to_remove(rng);
-
-        let mut active_peers = self.num_active_peers();
-        let mut failed_peers = self.num_failed_peers();
+    // Consider only the peers which the network views as being `Joined`, and check to see if a
+    // supermajority of these have polled blocks which allow the network to update its view of any
+    // `Joining` or `Leaving` peers, transitioning these to `Joined` or `Left` respectively.
+    pub fn update_network_views(all_peers: &mut BTreeMap<PeerId, Peer>) {
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply src/dev_utils/peer.rs]: I prefer the second suggestion; I'll use that.

File: src/dev_utils/peer.rs (hunk 11 above)
Comment (Fraser999) [reply src/dev_utils/peer.rs]: They are already, aren't they?

Comment (jeanphilippeD) [reply src/dev_utils/peer.rs]: I guess I was thinking above the loop, but that make sense here.
